File: 01.py
#!/usr/bin/env pypy3
# 1717/57

from util import *

if len(sys.argv) == 1:
    sys.stdin = open(__file__.replace("py", "in"))

res = 0

D = dict(zip("one, two, three, four, five, six, seven, eight, nine".split(", "), range(1, 10)))

for l in lines():
    digs = []
    for i, c in enumerate(l):
        if c.isdigit():
            digs.append(int(c))
        else:
            for k, v in D.items():
                if l[i:].startswith(k):
                    digs.append(v)
                    break
    # ls = list(map(int, (c for c in l if c.isdigit())))
    # if not ls: continue
    # print(ls)
    # assert ls[0] > 0 and ls[-1] > 0
    res += digs[0] * 10 + digs[-1]


prints(res)

File: 01.pl
main :-
    read_string(user_input, _, Input), split_string(Input, "\n", "\n", L),
    maplist(score, [part1, part2], [L, L], Ans),
    format("Part 1: ~d~nPart 2: ~d~n", Ans).

score(Part, Lines, Res) :-
    aggregate_all(sum(D1 * 10 + D2), (
        member(L, Lines),
        aggregate_all(r(min(I, D), max(I, D)), (
            valid_number(Part, Sub, D), sub_string(L, I, _, _, Sub)
        ), r(min(_, D1), max(_, D2)))
    ), Res).

valid_number(_, Sub, I) :- sub_string("_123456789", I, 1, _, Sub).
valid_number(part2, Sub, I) :-
    nth1(I, ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"], Sub).
